Mathlib/Topology/Algebra/Algebra.lean

```diff
@@ -536,6 +534,268 @@ end RestrictScalars
+class ContinuousAlgEquivClass (F : Type*) (R A B : outParam Type*) [CommSemiring R]
+    [Semiring A][TopologicalSpace A] [Semiring B] [TopologicalSpace B] [Algebra R A]
+    [Algebra R B] [EquivLike F A B] extends AlgEquivClass F R A B : Prop where
+  map_continuous : ∀ (f : F), Continuous f
+  inv_continuous : ∀ (f : F), Continuous (EquivLike.inv f)
+
+namespace ContinuousAlgEquiv
+
+variable {R A B C D : Type*}
+  [CommSemiring R] [Semiring A] [TopologicalSpace A] [Semiring B]
+  [TopologicalSpace B] [Semiring C] [TopologicalSpace C] [Semiring D] [TopologicalSpace D]
+  [Algebra R A] [Algebra R B] [Algebra R C] [Algebra R D]
+
+/-- A continuous algebra equivalence induces a continuous algebra map. -/
+@[coe]
+def toContinuousAlgHom (e : A ≃A[R] B) : A →A[R] B where
+  __ := e.toAlgHom
+  cont := e.continuous_toFun
+
+instance coe : Coe (A ≃A[R] B) (A →A[R] B) := ⟨toContinuousAlgHom⟩
+
+instance equivLike : EquivLike (A ≃A[R] B) A B where
+  coe f := f.toFun
+  inv f := f.invFun
+  coe_injective' f g h₁ h₂ := by
+    rcases f with ⟨⟨_, _⟩, _⟩
+    rcases g with ⟨⟨_, _⟩, _⟩
+    congr
+    exact Equiv.coe_fn_injective h₁
+  left_inv f := f.left_inv
+  right_inv f := f.right_inv
+
+instance continuousAlgEquivClass : ContinuousAlgEquivClass (A ≃A[R] B) R A B where
+  map_add f := f.map_add'
+  map_mul f := f.map_mul'
+  commutes f := f.commutes'
+  map_continuous := continuous_toFun
+  inv_continuous := continuous_invFun
+
+theorem coe_apply (e : A ≃A[R] B) (a : A) : (e : A →A[R] B) a = e a := rfl
+
+@[simp]
+theorem coe_coe (e : A ≃A[R] B) : ⇑(e : A →A[R] B) = e := rfl
+
+theorem toAlgEquiv_injective : Function.Injective (toAlgEquiv : (A ≃A[R] B) → A ≃ₐ[R] B) := by
+  rintro ⟨e, _, _⟩ ⟨e', _, _⟩ rfl
+  rfl
+
+@[ext]
+theorem ext {f g : A ≃A[R] B} (h : ⇑f = ⇑g) : f = g :=
+  toAlgEquiv_injective <| AlgEquiv.ext <| congr_fun h
+
+theorem coe_injective : Function.Injective ((↑) : (A ≃A[R] B) → A →A[R] B) :=
+  fun _ _ h => ext <| funext <| ContinuousAlgHom.ext_iff.1 h
+
+@[simp]
+theorem coe_inj {f g : A ≃A[R] B} : (f : A →A[R] B) = g ↔ f = g :=
+  coe_injective.eq_iff
+
+/-- A continuous algebra equivalence induces a homeomorphism. -/
+def toHomeomorph (e : A ≃A[R] B) : A ≃ₜ B where
+  __ := e
+  toEquiv := e.toAlgEquiv.toEquiv
+
+@[simp]
+theorem coe_toHomeomorph (e : A ≃A[R] B) : ⇑e.toHomeomorph = e := rfl
+
+theorem isOpenMap (e : A ≃A[R] B) : IsOpenMap e :=
+  e.toHomeomorph.isOpenMap
+
+theorem image_closure (e : A ≃A[R] B) (S : Set A) : e '' closure S = closure (e '' S) :=
+  e.toHomeomorph.image_closure S
+
+theorem preimage_closure (e : A ≃A[R] B) (S : Set B) : e ⁻¹' closure S = closure (e ⁻¹' S) :=
+  e.toHomeomorph.preimage_closure S
+
+@[simp]
+theorem isClosed_image (e : A ≃A[R] B) {S : Set A} : IsClosed (e '' S) ↔ IsClosed S :=
+  e.toHomeomorph.isClosed_image
+
+theorem map_nhds_eq (e : A ≃A[R] B) (a : A) : Filter.map e (𝓝 a) = 𝓝 (e a) :=
+  e.toHomeomorph.map_nhds_eq a
+
+theorem map_zero (e : A ≃A[R] B) : e (0 : A) = 0 :=
```

Comment:
You shouldn't need those; the root lemmas should work

Reply:
Removed those